### ISSUE TITLE
BUG: MPS is modified after ungrouping sites

### ISSUE THREAD

**Issue by flmachado** [tenpy/tenpy#197]
**Describe the bug**
When grouping and then ungrouping sites of an MPS, the MPS is modified as measured by local correlation functions.
Happy to provide the MPS if it helps track down the issue -- it has conservation of Sz and is in the cylinder.

**Minimal example or code extract**

```python
print("Initial Correlation Functions")
print(psi.correlation_function("Sz","Sz", sites1=range(3), sites2 = range(3)))
psi.canonical_form()

print("\n\nEnsuring they stay the same when caninalizating MPS")
print(psi.correlation_function("Sz","Sz", sites1=range(3), sites2 = range(3)))

# Grouping Sites
psi.group_sites(3)
print("\n\nCorrelation Function after grouping")
for i in range(3):
    for o in range(3):
        print(i, o, psi.correlation_function("Sz%d"%i,"Sz%d"%o, sites1=[0], sites2 = [0]).real)


# Ungrouping MPS -- initial bond dimension of the MPS is 512. Same result occurs if no trunc_params are provided.
psi.group_split({"chi_max":2048})

print("\n\nFinal Correlation Functions")
print(psi.correlation_function("Sz","Sz", sites1=range(3), sites2 = range(3)))
psi.canonical_form()

print("\n\Canonalizing MPS changes results")
print(psi.correlation_function("Sz","Sz", sites1=range(3), sites2 = range(3)))
```

Which has the following output
```
Initial Correlation Functions
[[ 0.25    -0.10817 -0.03445]
 [-0.10817  0.25    -0.10736]
 [-0.03445 -0.10736  0.25   ]]


Ensuring they stay the same when caninalizating MPS
[[ 0.25    -0.10817 -0.03445]
 [-0.10817  0.25    -0.10736]
 [-0.03445 -0.10736  0.25   ]]


Correlation Function after grouping
0 0 [[0.25]]
0 1 [[-0.10817]]
0 2 [[-0.03445]]
1 0 [[-0.10817]]
1 1 [[0.25]]
1 2 [[-0.10736]]
2 0 [[-0.03445]]
2 1 [[-0.10736]]
2 2 [[0.25]]
INFO    : truncation: reading 'chi_max'=2048
INFO    : truncation: reading 'chi_max'=2048
INFO    : truncation: reading 'chi_min'=None
INFO    : truncation: reading 'degeneracy_tol'=None
INFO    : truncation: reading 'svd_min'=1e-14
INFO    : truncation: reading 'trunc_cut'=1e-14
INFO    : truncation: reading 'chi_max'=2048
INFO    : truncation: reading 'chi_min'=None
INFO    : truncation: reading 'degeneracy_tol'=None
INFO    : truncation: reading 'svd_min'=1e-14
INFO    : truncation: reading 'trunc_cut'=1e-14
INFO    : truncation: reading 'chi_max'=2048
INFO    : truncation: reading 'chi_min'=None
INFO    : truncation: reading 'degeneracy_tol'=None
INFO    : truncation: reading 'svd_min'=1e-14
INFO    : truncation: reading 'trunc_cut'=1e-14
INFO    : truncation: reading 'chi_max'=2048
INFO    : truncation: reading 'chi_min'=None
INFO    : truncation: reading 'degeneracy_tol'=None
INFO    : truncation: reading 'svd_min'=1e-14
INFO    : truncation: reading 'trunc_cut'=1e-14
INFO    : truncation: reading 'chi_max'=2048
INFO    : truncation: reading 'chi_min'=None
INFO    : truncation: reading 'degeneracy_tol'=None
INFO    : truncation: reading 'svd_min'=1e-14
INFO    : truncation: reading 'trunc_cut'=1e-14
INFO    : truncation: reading 'chi_max'=2048
INFO    : truncation: reading 'chi_min'=None
INFO    : truncation: reading 'degeneracy_tol'=None
INFO    : truncation: reading 'svd_min'=1e-14
INFO    : truncation: reading 'trunc_cut'=1e-14
INFO    : truncation: reading 'chi_max'=2048
INFO    : truncation: reading 'chi_min'=None
INFO    : truncation: reading 'degeneracy_tol'=None
INFO    : truncation: reading 'svd_min'=1e-14
INFO    : truncation: reading 'trunc_cut'=1e-14


Final Correlation Functions
[[ 0.2174  -0.00039 -0.07632]
 [-0.00039  0.25    -0.10529]
 [-0.07632 -0.10529  0.25   ]]

\Canonalizing MPS changes results
[[ 0.25    -0.02045 -0.07302]
 [-0.02045  0.25    -0.12983]
 [-0.07302 -0.12983  0.25   ]]


```

Current config
```
tenpy 0.9.0.dev160+ba4899b (not compiled),
git revision ba4899b44610a5a923e2dc0e9468a07296c5543a using
python 3.9.7 (default, Sep 16 2021, 13:09:58) 
[GCC 7.5.0]
numpy 1.20.3, scipy 1.7.3
```

**Comment by Jakob-Unfried** [comment]
thanks for submitting the issue.
yes, could you provide an MPS that exhibits the buggy behavior?
if you have code that generates such an MPS, that would be even better

**Comment by flmachado** [comment]
Hi Jakob,

Any state seems to exhibit this behavior.
Below I prepare a random initial state and observe the same behavior.


```
from tenpy.networks.site import SpinHalfSite

psi = tenpy.networks.mps.MPS.from_product_state(sites = [SpinHalfSite() for i in range(12)], 
                                          p_state = ["up", "down"]*6)

from tenpy.algorithms.tebd import RandomUnitaryEvolution

sim = RandomUnitaryEvolution(psi, options = {"N_steps" : 4})
sim.run()

print("Initial Correlation Functions")
print(psi.correlation_function("Sz","Sz")[:5,:5] )
psi.canonical_form()

print("\n\nEnsuring they stay the same when caninalizating MPS")
print(psi.correlation_function("Sz","Sz")[:5,:5] )

# Grouping Sites
psi.group_sites(3)
print("\n\nCorrelation Function after grouping")
for i in range(3):
    for o in range(3):
        print(i, o, psi.correlation_function("Sz%d"%i,"Sz%d"%o, sites1=[0], sites2 = [0]).real)


# Ungrouping MPS -- initial bond dimension of the MPS is 512. Same result occurs if no trunc_params are provided.
psi.group_split({"chi_max":2048})

print("\n\nFinal Correlation Functions")
print(psi.correlation_function("Sz","Sz")[:5,:5])
psi.canonical_form()

print("\n\nCanonalizing MPS changes results")
print(psi.correlation_function("Sz","Sz")[:5,:5])
```


Here's the output I get (including some info):
```
INFO:tenpy.tools.params:RandomUnitaryEvolution: reading 'N_steps'=4
INFO:tenpy.algorithms.tebd:--> time=4.000, max(chi)=64, max(S)=1.93635, avg DeltaS=1.4213e+00, since last update: 0.1s

Initial Correlation Functions
[[ 0.25       -0.09569012 -0.1183616  -0.01478253 -0.03703739]
 [-0.09569012  0.25        0.06647626 -0.00393626 -0.0567045 ]
 [-0.1183616   0.06647626  0.25       -0.04659406 -0.02473925]
 [-0.01478253 -0.00393626 -0.04659406  0.25       -0.08304205]
 [-0.03703739 -0.0567045  -0.02473925 -0.08304205  0.25      ]]


Ensuring they stay the same when caninalizating MPS

INFO:tenpy.tools.params:truncation: reading 'chi_max'=2048
INFO:tenpy.tools.params:truncation: reading 'chi_max'=2048
INFO:tenpy.tools.params:truncation: reading 'chi_min'=None
INFO:tenpy.tools.params:truncation: reading 'degeneracy_tol'=None
INFO:tenpy.tools.params:truncation: reading 'svd_min'=1e-14
INFO:tenpy.tools.params:truncation: reading 'trunc_cut'=1e-14
INFO:tenpy.tools.params:truncation: reading 'chi_max'=2048
INFO:tenpy.tools.params:truncation: reading 'chi_min'=None
INFO:tenpy.tools.params:truncation: reading 'degeneracy_tol'=None
INFO:tenpy.tools.params:truncation: reading 'svd_min'=1e-14
INFO:tenpy.tools.params:truncation: reading 'trunc_cut'=1e-14
INFO:tenpy.tools.params:truncation: reading 'chi_max'=2048
INFO:tenpy.tools.params:truncation: reading 'chi_min'=None
INFO:tenpy.tools.params:truncation: reading 'degeneracy_tol'=None
INFO:tenpy.tools.params:truncation: reading 'svd_min'=1e-14
INFO:tenpy.tools.params:truncation: reading 'trunc_cut'=1e-14
INFO:tenpy.tools.params:truncation: reading 'chi_max'=2048
INFO:tenpy.tools.params:truncation: reading 'chi_min'=None
INFO:tenpy.tools.params:truncation: reading 'degeneracy_tol'=None
INFO:tenpy.tools.params:truncation: reading 'svd_min'=1e-14
INFO:tenpy.tools.params:truncation: reading 'trunc_cut'=1e-14
INFO:tenpy.tools.params:truncation: reading 'chi_max'=2048
INFO:tenpy.tools.params:truncation: reading 'chi_min'=None
INFO:tenpy.tools.params:truncation: reading 'degeneracy_tol'=None
INFO:tenpy.tools.params:truncation: reading 'svd_min'=1e-14
INFO:tenpy.tools.params:truncation: reading 'trunc_cut'=1e-14
INFO:tenpy.tools.params:truncation: reading 'chi_max'=2048
INFO:tenpy.tools.params:truncation: reading 'chi_min'=None
INFO:tenpy.tools.params:truncation: reading 'degeneracy_tol'=None
INFO:tenpy.tools.params:truncation: reading 'svd_min'=1e-14
INFO:tenpy.tools.params:truncation: reading 'trunc_cut'=1e-14
INFO:tenpy.tools.params:truncation: reading 'chi_max'=2048
INFO:tenpy.tools.params:truncation: reading 'chi_min'=None
INFO:tenpy.tools.params:truncation: reading 'degeneracy_tol'=None
INFO:tenpy.tools.params:truncation: reading 'svd_min'=1e-14
INFO:tenpy.tools.params:truncation: reading 'trunc_cut'=1e-14

[[ 0.25       -0.09569012 -0.1183616  -0.01478253 -0.03703739]
 [-0.09569012  0.25        0.06647626 -0.00393626 -0.0567045 ]
 [-0.1183616   0.06647626  0.25       -0.04659406 -0.02473925]
 [-0.01478253 -0.00393626 -0.04659406  0.25       -0.08304205]
 [-0.03703739 -0.0567045  -0.02473925 -0.08304205  0.25      ]]


Correlation Function after grouping
0 0 [[0.25]]
0 1 [[-0.09569012]]
0 2 [[-0.1183616]]
1 0 [[-0.09569012]]
1 1 [[0.25]]
1 2 [[0.06647626]]
2 0 [[-0.1183616]]
2 1 [[0.06647626]]
2 2 [[0.25]]


Final Correlation Functions
[[ 2.36374355e-01 -1.11022302e-16 -1.11228230e-01  2.00912191e-03
   4.98732999e-18]
 [-1.11022302e-16  2.50000000e-01  1.52095796e-02  6.74804033e-03
   7.80625564e-18]
 [-1.11228230e-01  1.52095796e-02  2.50000000e-01 -2.61321021e-02
   1.84314369e-18]
 [ 2.00912191e-03  6.74804033e-03 -2.61321021e-02  1.12893303e-01
   2.00577402e-17]
 [ 4.98732999e-18  7.80625564e-18  1.84314369e-18  2.00577402e-17
   2.50000000e-01]]


Canonalizing MPS changes results
[[ 2.50000000e-01 -3.30529243e-02 -1.47136650e-01  4.88487473e-03
  -2.16840434e-17]
 [-3.30529243e-02  2.50000000e-01  1.55300504e-02  1.71914403e-02
  -1.12757026e-17]
 [-1.47136650e-01  1.55300504e-02  2.50000000e-01 -5.82822613e-02
   1.82145965e-17]
 [ 4.88487473e-03  1.71914403e-02 -5.82822613e-02  2.50000000e-01
   7.54604712e-17]
 [-2.16840434e-17 -1.12757026e-17  1.82145965e-17  7.54604712e-17
   2.50000000e-01]]
```

**Comment by Jakob-Unfried** [comment]
I cant reproduce your behavior, see the test added in PR #207.

I also ran your snippet with minimal modifications:

```python
def test_197_flmachado():
    psi = mps.MPS.from_product_state(sites = [site.SpinHalfSite() for i in range(12)],
                                     p_state = ["up", "down"]*6)
    sim = tebd.RandomUnitaryEvolution(psi, options = {"N_steps" : 4})
    sim.run()
    print("Initial Correlation Functions")
    corr_init = psi.correlation_function("Sz","Sz")
    print(corr_init[:5,:5])
    psi.canonical_form()

    print("\n\nEnsuring they stay the same when caninalizating MPS")
    print(psi.correlation_function("Sz","Sz")[:5,:5] )

    # Grouping Sites
    psi.group_sites(3)
    print("\n\nCorrelation Function after grouping")
    for i in range(3):
        for o in range(3):
            print(i, o, psi.correlation_function("Sz%d"%i,"Sz%d"%o, sites1=[0], sites2 = [0]).real)


    # Ungrouping MPS -- initial bond dimension of the MPS is 512. Same result occurs if no trunc_params are provided.
    psi.group_split({"chi_max":2048})

    print("\n\nFinal Correlation Functions")
    corr_split = psi.correlation_function("Sz","Sz")
    print(corr_split[:5,:5])
    assert np.allclose(corr_split, corr_init)

    psi.canonical_form()
    print("\n\nCanonalizing MPS changes results")
    corr_split_canonical = psi.correlation_function("Sz","Sz")
    print(corr_split_canonical[:5,:5])
    assert np.allclose(corr_split_canonical, corr_init)

    assert False  # debugging: causes pytest to show the printed output
```

and got correct outputs

```
Initial Correlation Functions
[[ 0.25        0.00639754 -0.13805102  0.02011331 -0.06859113]
 [ 0.00639754  0.25       -0.04694951 -0.04072142 -0.04398974]
 [-0.13805102 -0.04694951  0.25       -0.0877797   0.00765237]
 [ 0.02011331 -0.04072142 -0.0877797   0.25       -0.06848878]
 [-0.06859113 -0.04398974  0.00765237 -0.06848878  0.25      ]]


Ensuring they stay the same when caninalizating MPS
[[ 0.25        0.00639754 -0.13805102  0.02011331 -0.06859113]
 [ 0.00639754  0.25       -0.04694951 -0.04072142 -0.04398974]
 [-0.13805102 -0.04694951  0.25       -0.0877797   0.00765237]
 [ 0.02011331 -0.04072142 -0.0877797   0.25       -0.06848878]
 [-0.06859113 -0.04398974  0.00765237 -0.06848878  0.25      ]]


Correlation Function after grouping
0 0 [[0.25]]
0 1 [[0.00639754]]
0 2 [[-0.13805102]]
1 0 [[0.00639754]]
1 1 [[0.25]]
1 2 [[-0.04694951]]
2 0 [[-0.13805102]]
2 1 [[-0.04694951]]
2 2 [[0.25]]


Final Correlation Functions
[[ 0.25        0.00639754 -0.13805102  0.02011331 -0.06859113]
 [ 0.00639754  0.25       -0.04694951 -0.04072142 -0.04398974]
 [-0.13805102 -0.04694951  0.25       -0.0877797   0.00765237]
 [ 0.02011331 -0.04072142 -0.0877797   0.25       -0.06848878]
 [-0.06859113 -0.04398974  0.00765237 -0.06848878  0.25      ]]


Canonalizing MPS changes results
[[ 0.25        0.00639754 -0.13805102  0.02011331 -0.06859113]
 [ 0.00639754  0.25       -0.04694951 -0.04072142 -0.04398974]
 [-0.13805102 -0.04694951  0.25       -0.0877797   0.00765237]
 [ 0.02011331 -0.04072142 -0.0877797   0.25       -0.06848878]
 [-0.06859113 -0.04398974  0.00765237 -0.06848878  0.25      ]]
```

**Comment by Jakob-Unfried** [comment]
what version of tenpy are you running?
I.e. could you show me the output of `tenpy.show_config()`?

**Comment by Jakob-Unfried** [comment]
I do get the buggy behavior when i checkout the tenpy commit you are on.
It seems to have been fixed already, possibly by [this commit](https://github.com/tenpy/tenpy/commit/75ec249ec899e0bec7d32b7b91f7f4310a8f7ce0).

Looks like you just need to update tenpy.
Since the above commit is not yet included in the 0.10 release, you will need to [install from source](https://tenpy.readthedocs.io/en/latest/install/from_source.html) if you want a fixed version in the near future.

If that does not help you, feel free to re-open this or re-submit a new issue.
